File: src/main/sparkcore/test/myself/PairRddPractice.scala
package test.myself

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
  * PairRDD练习
  */
object PairRddPractice {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
    conf.setAppName("PairRddPractice")
    conf.setMaster("local")

    val sc = new SparkContext(conf)
    sc.setLogLevel("Error")

    val data = List((100, "spark"), (1112, "hadoop"), (1113, "hive"), (49, "hbase"), (15, "kafka"), (10, "stream"))

    val unit = sc.parallelize(data)
    unit.foreach(x => println(x._1 + "--" + x._2))

    val d1 = List("spark hadoop", "hive hbase", "kafka stream")
    val unit1 = sc.parallelize(d1)
    unit1.map(x => (x.split(" ")(0), x.split(" ")(1)))
      .foreach(line => println(line._1 + " " + line._2))

    unit.filter(x => x._2.equals("spark"))
      .map(y => y._1)
      .foreach(println)

    reduceByKey(sc)

    sortKey(unit, sc)

    join(sc)

    combineBy(sc)

    partitions(sc)
  }

  /**
    * spark版的wordcount功能
    *
    */
  def reduceByKey(sc: SparkContext): Unit = {
    val value = sc.textFile("in/test.csv")

    value.flatMap(x => x.split(","))
      .map(x => (x, 1))
      .reduceByKey((x, y) => x + y)
      .foreach(println)
  }

  /**
    * 当我们在(K, V)数据集中应用sortByKey()函数时，数据是根据RDD中的键K排序的
    *
    * @param sc
    */
  def sortKey(rdd: RDD[(Int, String)], sc: SparkContext): Unit = {
    println("---------------------------------------------sortKey---------------------------------------------")
    rdd.sortByKey(false)
      .foreach(println)
  }

  /**
    * join是数据库术语。它使用公共值组合两个表中的字段。Spark中的join()操作是在pairRDD上定义的。
    * pairRDD每个元素都以tuple的形式出现。
    * tuple第一个元素是key，第二个元素是value。join()操作根据key组合两个数据集。
    *
    * 注意:join()根据key连接两个不同的RDDs。
    *
    * @param sc
    */
  def join(sc: SparkContext): Unit = {
    println("-------------------------------------------------join---------------------------------------------")
    val data = sc.parallelize(Array(('A', 1), ('b', 2)))
    val data2 = sc.parallelize(Array(('A', 4), ('A', 6), ('b', 7), ('c', 3), ('c', 8)))

    data.join(data2)
      .collect()
      .foreach(println)
  }

  case class ScoreDetail(name: String, subject: String, score: Float)

  /**
    *
    *
    * createCombiner为V => Iterable[V]。
    * mergeValue则将原RDD中Pair的Value合并为操作后的C类型数据。合并操作的实现决定了结果的运算方式。所以，mergeValue更像是声明了一种合并方式，
    * 它是由整个combine运算的结果来导向的。
    * 函数的输入为原RDD中Pair的V，输出为结果RDD中Pair的C。
    * 最后的mergeCombiners则会根据每个Key对应的多个C，进行归并。
    *
    * @param sc
    */
  def combineBy(sc: SparkContext): Unit = {
    val scores = List(
      ScoreDetail("lili", "Math", 98),
      ScoreDetail("lili", "English", 88),
      ScoreDetail("zhangsan", "Math", 75),
      ScoreDetail("zhangsan", "English", 78),
      ScoreDetail("wangliu", "Math", 90),
      ScoreDetail("wangliu", "English", 80),
      ScoreDetail("lisi", "Math", 91),
      ScoreDetail("lisi", "English", 80)
    )

    //生成ScoreDetail集合
    val sd = for (i <- scores) yield (i.name, i)
    //("lili",ScoreDetail)
    //转换成PairRdd
    val pad = sc.parallelize(sd)
    println("--------------------学生的个人平均分-----------------------")
    //计算平均分数
    pad.combineByKey(
      (x: ScoreDetail) => (x.score, 1), //(98,1)
      (acc: (Float, Int), x: ScoreDetail) => (acc._1 + x.score, acc._2 + 1), //(总分，科目数)
      (acc1: (Float, Int), acc2: (Float, Int)) => (acc1._1 + acc2._1, acc1._2 + acc2._2)
    )
      .map({
        case (key, value) => (key, value._1 / value._2)
      })
      .collect()
      .foreach(println)

    val rdd = sc.parallelize(for (i <- scores) yield (i.subject, i))

    println("----------------单科成绩平均分-----------------")
    //计算学科的平均分
    rdd.combineByKey(
      (x: ScoreDetail) => (x.score, 1),
      (score: (Float, Int), s: ScoreDetail) => (score._1 + s.score, score._2 + 1),
      (acc1: (Float, Int), acc2: (Float, Int)) => (acc1._1 + acc2._1, acc1._2 + acc2._2)
    )
      .map({
        case (k, v) => (k, v._1 / v._2)
      })
      .collect()
      .foreach(println)
  }

  /**
    *
    * @param sc
    */
  def partitions(sc: SparkContext): Unit = {
    val rdd1 = sc.parallelize(Seq((1, "jan", 2016), (3, "nov", 2014), (16, "feb", 2014)), 1)
    val rdd2 = sc.parallelize(Seq((5, "dec", 2014), (17, "sep", 2015)), 1)
    val rdd3 = sc.parallelize(Seq((6, "dec", 2011), (16, "may", 2015)), 1)
    val rddUnion = rdd1.union(rdd2).union(rdd3)
    println("rddUnion.partitions.size:" + rddUnion.partitions.size)
    rddUnion.foreach(println)
  }

}

File: src/main/java/sparkcorejava/test/rdd/PairRddPractice.java
package sparkcorejava.test.rdd;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import org.apache.spark.storage.StorageLevel;
import scala.Tuple2;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * pairRdd 练习
 */
public class PairRddPractice {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName(PairRddPractice.class.getSimpleName()).setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);
        sc.setLogLevel("ERROR");

//        createPairRdd(sc);
//
//        JavaRDD<String> rdd = sc.textFile("in/test.csv");
        JavaRDD<String> rdd = sc.textFile("in/student.csv");
//        createPairRdd(rdd);
//        normalToPairRdd(sc);
        reduceByKey(rdd);
//        average(sc);
//        combineByKey(sc);
//        averageScore(rdd);

//        join(sc);

//        sc.stop();
    }

    /**
     * 通过外部数据源
     *
     * @param rdd
     */
    private static void createPairRdd(JavaRDD<String> rdd) {
        rdd.foreach(new VoidFunction<String>() {
            @Override
            public void call(String s) throws Exception {
                System.out.println("createPairRdd result is " + s);
            }
        });
    }

    /**
     * 使用集合创建RDD
     *
     * @param sc
     */
    private static void createPairRdd(JavaSparkContext sc) {
        List<Tuple2<String, Integer>> tuple2s = Arrays.asList(new Tuple2<>("liyang", 22), new Tuple2<>("lili", 21),
                new Tuple2<>("zhaosi", 22), new Tuple2<>("zhaowu", 22), new Tuple2<>("zhouyi", 23), new Tuple2<>("zhoubing", 22),
                new Tuple2<>("cuixinle", 22), new Tuple2<>("caiquanwu", 22), new Tuple2<>("wudajing", 22), new Tuple2<>("guli", 23), new Tuple2<>("guosan", 21));

        sc.parallelize(tuple2s).foreach(new VoidFunction<Tuple2<String, Integer>>() {
            @Override
            public void call(Tuple2<String, Integer> t1) throws Exception {
                System.out.println("createPairRdd result is " + t1);
            }
        });
    }

    private static void normalToPairRdd(JavaSparkContext sc) {
        if (sc == null) {
            return;
        }
        List<String> list = Arrays.asList("liyang 200", "lili 30", "lili 21", "susi 25");

        JavaRDD<String> stringRdd = sc.parallelize(list);
        stringRdd.persist(StorageLevel.MEMORY_ONLY());
//        JavaPairRDD<String, Integer> pairRdd = stringRdd.mapToPair(new PairFunction<String, String, Integer>() {
//            @Override
//            public Tuple2<String, Integer> call(String s) throws Exception {
//                return new Tuple2<String, Integer>(s.split(" ")[0], Integer.valueOf(s.split(" ")[1]));
//            }
//        });
//
//        pairRdd.reduceByKey(new Function2<Integer, Integer, Integer>() {
//            @Override
//            public Integer call(Integer v1, Integer v2) throws Exception {
//                return v1 + v2;
//            }
//        }).foreach(new VoidFunction<Tuple2<String, Integer>>() {
//            @Override
//            public void call(Tuple2<String, Integer> tuple2) throws Exception {
//                System.out.println("normalToPairRdd result is " + tuple2);
//            }
//        });

        stringRdd.mapToPair(new PairFunction<String, Integer, String>() {

            @Override
            public Tuple2<Integer, String> call(String s) throws Exception {
                return new Tuple2<Integer, String>(Integer.valueOf(s.split(" ")[1]), s.split(" ")[0]);
            }
        }).filter(new Function<Tuple2<Integer, String>, Boolean>() {
            @Override
            public Boolean call(Tuple2<Integer, String> v1) throws Exception {
                return v1._1 > 20;
            }
        }).reduceByKey(new Function2<String, String, String>() {
            @Override
            public String call(String v1, String v2) throws Exception {
                return v1;
            }
        }).foreach(new VoidFunction<Tuple2<Integer, String>>() {
            @Override
            public void call(Tuple2<Integer, String> integerStringTuple2) throws Exception {
                System.out.println(" result is " + integerStringTuple2);
            }
        });

    }

    private static void reduceByKey(JavaRDD<String> rdd) {
        if (rdd == null) {
            return;
        }
        rdd.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String s) throws Exception {
                String[] split = s.split(",");
                return Arrays.asList(split).iterator();
            }
        }).mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        }).sortByKey()
                .foreach(new VoidFunction<Tuple2<String, Integer>>() {
                    @Override
                    public void call(Tuple2<String, Integer> tuple2) throws Exception {
                        System.out.println(" reduceByKey:(" + tuple2._1 + ":\t" + tuple2._2 + ")");
                    }
                });
    }

    private static void averageScore(JavaRDD<String> rdd) {
        if (rdd == null) {
            return;
        }

        JavaPairRDD<String, ScoreDetail> pair = rdd.map(new Function<String, ScoreDetail>() {
            @Override
            public ScoreDetail call(String v1) throws Exception {
                String[] split = v1.split(",");
                ScoreDetail detail = new ScoreDetail();
                detail.setName(split[0]);
                detail.setSubject(split[1]);
                detail.setScore(Integer.valueOf(split[2]));
                return detail;
            }
        }).mapToPair(new PairFunction<ScoreDetail, String, ScoreDetail>() {
            @Override
            public Tuple2<String, ScoreDetail> call(ScoreDetail scoreDetail) throws Exception {
                return new Tuple2<String, ScoreDetail>(scoreDetail.getName(), scoreDetail);
            }
        });
        JavaPairRDD<String, ScoreDetail> reduce = pair.reduceByKey(new Function2<ScoreDetail, ScoreDetail, ScoreDetail>() {
            @Override
            public ScoreDetail call(ScoreDetail v1, ScoreDetail v2) throws Exception {
                return new ScoreDetail(v1.getName(), v1.getSubject() + "," + v2.getSubject(), v1.getScore() + v2.getScore());
            }
        });
        reduce.foreach(new VoidFunction<Tuple2<String, ScoreDetail>>() {
            @Override
            public void call(Tuple2<String, ScoreDetail> t) throws Exception {
                System.out.println(t._1 + ": " + t._2.getScore() / t._2.getSubject().split(",").length);
            }
        });
    }

    public static JavaPairRDD<String, Tuple2<String, String>> join(JavaSparkContext sc) {
        if (sc == null) {
            return null;
        }
        JavaRDD<String> rdd1 = sc.textFile("in/test.csv");

        JavaPairRDD<String, String> fire = rdd1.filter(new Function<String, Boolean>() {
            @Override
            public Boolean call(String v1) throws Exception {
                return v1.contains("FIRE");
            }
        }).mapToPair((PairFunction<String, String, String>) s -> {
            String[] split = s.split(",");
            return new Tuple2<>(split[0], split[3]);
        });

        JavaPairRDD<String, String> lieutenant = rdd1.filter(new Function<String, Boolean>() {
            @Override
            public Boolean call(String v1) throws Exception {
                return v1.contains("LIEUTENANT");
            }
        }).mapToPair(new PairFunction<String, String, String>() {
            @Override
            public Tuple2<String, String> call(String s) throws Exception {
                String[] split = s.split(",");
                return new Tuple2<>(split[0], split[2]);
            }
        });

        JavaPairRDD<String, Tuple2<String, String>> join = fire.join(lieutenant);
//        join.sortByKey().foreach(new VoidFunction<Tuple2<String, Tuple2<String, String>>>() {
//            @Override
//            public void call(Tuple2<String, Tuple2<String, String>> tuple2) throws Exception {
//                System.out.println(" join result is : " + tuple2);
//            }
//        });
        return join;
    }

    private static void average(JavaSparkContext sc) {
        if (sc == null) {
            return;
        }
        List<Tuple2<String, Double>> tuple2s = Arrays.asList(new Tuple2<>("Fred", 88.0), new Tuple2<>("Fred", 95.0), new Tuple2<>("Fred", 91.0),
                new Tuple2<>("Wilma", 93.0), new Tuple2<>("Wilma", 95.0), new Tuple2<>("Wilma", 98.0));

        //rdd(学生，成绩)
        JavaPairRDD<String, Double> rdd = sc.parallelizePairs(tuple2s);

        //Function(V,C)（成绩，科目数）
        //例如 此时返回的是（88.0，1） 成绩是88 科目数为1
        Function<Double, Tuple2<Double, Integer>> createCombiner = new Function<Double, Tuple2<Double, Integer>>() {

            @Override
            public Tuple2<Double, Integer> call(Double v1) throws Exception {
                return new Tuple2<>(v1, 1);
            }
        };
        // //Function2(C,V,C)（成绩，（成绩，科目数））
        /**
         * 这里的v1  得到的就是（88.0，1）在同一个分区内 我们又碰到了一个Fred 所以需要将之前的分数+现在的这个分数：v1._1 + v2,科目数量+1 即 v1._2 + 1
         */
        Function2<Tuple2<Double, Integer>, Double, Tuple2<Double, Integer>> mergeValue = new Function2<Tuple2<Double, Integer>, Double, Tuple2<Double, Integer>>() {

            @Override
            public Tuple2<Double, Integer> call(Tuple2<Double, Integer> v1, Double v2) throws Exception {
                return new Tuple2<Double, Integer>(v1._1 + v2, v1._2 + 1);
            }
        };
        /**
         * Fred 的数据可能分布在多个分区，因此需要进行合并，及多个分区的同个人的成绩相加：v1._1 + v2._1，科目相加 v1._2 + v2._2
         */
        Function2<Tuple2<Double, Integer>, Tuple2<Double, Integer>, Tuple2<Double, Integer>> mergeCombiners = new Function2<Tuple2<Double, Integer>, Tuple2<Double, Integer>, Tuple2<Double, Integer>>() {

            @Override
            public Tuple2<Double, Integer> call(Tuple2<Double, Integer> v1, Tuple2<Double, Integer> v2) throws Exception {
                return new Tuple2<>(v1._1 + v2._1, v1._2 + v2._2);
            }
        };
        //聚合处理 (学生,(总分，总科目))
        rdd.combineByKey(createCombiner, mergeValue, mergeCombiners).map(new Function<Tuple2<String, Tuple2<Double, Integer>>, Tuple2<String, Double>>() {
            @Override
            public Tuple2<String, Double> call(Tuple2<String, Tuple2<Double, Integer>> v1) throws Exception {
                return new Tuple2<String, Double>(v1._1, v1._2._1 / v1._2._2);
            }
        }).foreach(new VoidFunction<Tuple2<String, Double>>() {
            @Override
            public void call(Tuple2<String, Double> aDouble) throws Exception {
                System.out.println(aDouble._1 + ":\t" + aDouble._2());
            }
        });

    }

    /**
     * createCombiner: V => C ，这个函数把当前的值作为参数，此时我们可以对其做些附加操作(类型转换)并把它返回 (这一步类似于初始化操作)
     * mergeValue: (C, V) => C，该函数把元素V合并到之前的元素C(createCombiner)上 (这个操作在每个分区内进行)
     * mergeCombiners: (C, C) => C，该函数把2个元素C合并 (这个操作在不同分区间进行)
     * <p>
     * combineByKey(Function(V,C),Function2(C,V,C),Function2(C,C,C)）
     *
     * @param sc
     */
    private static void combineByKey(JavaSparkContext sc) {
        if (sc == null) {
            return;
        }
        JavaRDD<String> rdd = sc.textFile("in/student.csv");

        JavaRDD<ScoreDetail> student = rdd.map(new Function<String, ScoreDetail>() {
            @Override
            public ScoreDetail call(String v1) throws Exception {
                String[] split = v1.split(",");
                ScoreDetail detail = new ScoreDetail();
                detail.setName(split[0]);
                detail.setSubject(split[1]);
                detail.setScore(Integer.valueOf(split[2]));
                return detail;
            }
        });
        JavaPairRDD<String, ScoreDetail> pair = student.mapToPair(new PairFunction<ScoreDetail, String, ScoreDetail>() {
            @Override
            public Tuple2<String, ScoreDetail> call(ScoreDetail scoreDetail) throws Exception {
                return new Tuple2<String, ScoreDetail>(scoreDetail.getName(), scoreDetail);
            }
        });

        JavaPairRDD<String, ScoreDetail> subject = student.mapToPair(new PairFunction<ScoreDetail, String, ScoreDetail>() {
            @Override
            public Tuple2<String, ScoreDetail> call(ScoreDetail scoreDetail) throws Exception {
                return new Tuple2<String, ScoreDetail>(scoreDetail.getSubject(), scoreDetail);
            }
        });

        //1、创建createCombiner：组合器函数，输入参数为RDD[K,V]中的V（即ScoreDetail对象），输出为tuple2(学生成绩，1)
        Function<ScoreDetail, Tuple2<Float, Integer>> createCombiner = new Function<ScoreDetail, Tuple2<Float, Integer>>() {
            @Override
            public Tuple2<Float, Integer> call(ScoreDetail v1) throws Exception {
                return new Tuple2<Float, Integer>((float) v1.score, 1);
            }
        };

        //2、mergeValue：合并值函数，输入参数为(C,V)即（tuple2（学生成绩，1），ScoreDetail对象），输出为tuple2（学生成绩，2）
        Function2<Tuple2<Float, Integer>, ScoreDetail, Tuple2<Float, Integer>> mergeValue = new Function2<Tuple2<Float, Integer>, ScoreDetail, Tuple2<Float, Integer>>() {
            @Override
            public Tuple2<Float, Integer> call(Tuple2<Float, Integer> v1, ScoreDetail v2) throws Exception {
                return new Tuple2<Float, Integer>(v1._1() + v2.score, v1._2() + 1);
            }
        };

        //3、mergeCombiners：合并组合器函数，对多个节点上的数据合并，输入参数为(C,C)，输出为C

        Function2<Tuple2<Float, Integer>, Tuple2<Float, Integer>, Tuple2<Float, Integer>> mergeCombiners = new Function2<Tuple2<Float, Integer>, Tuple2<Float, Integer>, Tuple2<Float, Integer>>() {
            @Override
            public Tuple2<Float, Integer> call(Tuple2<Float, Integer> v1, Tuple2<Float, Integer> v2) throws Exception {
                return new Tuple2<Float, Integer>(v1._1() + v2._1(), v1._2() + v2._2());
            }
        };

        //4、combineByKey并求均值
        JavaPairRDD<String, Float> res = pair.combineByKey(createCombiner, mergeValue, mergeCombiners, 2)
                .mapToPair(x -> new Tuple2<String, Float>(x._1(), x._2()._1() / x._2()._2()));
        res.foreach(new VoidFunction<Tuple2<String, Float>>() {
            @Override
            public void call(Tuple2<String, Float> tuple2) throws Exception {
                System.out.println(" student average:(" + tuple2._1 + ":\t" + tuple2._2 + ")");
            }
        });
        subject.combineByKey(createCombiner, mergeValue, mergeCombiners, 2)
                .mapToPair(new PairFunction<Tuple2<String, Tuple2<Float, Integer>>, String, Float>() {
                    @Override
                    public Tuple2<String, Float> call(Tuple2<String, Tuple2<Float, Integer>> tuple2) throws Exception {
                        return new Tuple2<String, Float>(tuple2._1, tuple2._2._1 / tuple2._2._2);
                    }
                })
                .foreach(new VoidFunction<Tuple2<String, Float>>() {
                    @Override
                    public void call(Tuple2<String, Float> tuple2) throws Exception {
                        System.out.println(" subject average:(" + tuple2._1 + ":\t" + tuple2._2 + ")");
                    }
                });
    }

    public static class ScoreDetail implements Serializable {
        String name;
        String subject;
        int score;

        public ScoreDetail() {
        }

        public ScoreDetail(String name, String subject, int score) {
            this.name = name;
            this.subject = subject;
            this.score = score;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getSubject() {
            return subject;
        }

        public void setSubject(String subject) {
            this.subject = subject;
        }

        public int getScore() {
            return score;
        }

        public void setScore(int score) {
            this.score = score;
        }

        @Override
        public String toString() {
            return super.toString();
        }
    }

}
